connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala

```diff
@@ -386,143 +365,152 @@ object KustoWriter {
-            kustoClient)
-          curBlobUUID = UUID.randomUUID().toString
-          createBlobWriter(
-            parameters.coordinates,
-            parameters.tmpTableName,
-            kustoClient,
-            partitionIdString,
-            row._2,
-            curBlobUUID)
-        }
+    val lastBlobWriter = rows.zipWithIndex.foldLeft[BlobWriteResource](initialBlobWriter) { case (blobWriter, row) =>
+      RowCSVWriterUtils.writeRowAsCSV(row._1, parameters.schema, timeZone, blobWriter.csvWriter)
+      val count = blobWriter.csvWriter.getCounter
+      val shouldNotCommitBlockBlob = count < maxBlobSize
+      if (shouldNotCommitBlockBlob) {
+        blobWriter
+      } else {
+        val blobIndexInBatch = blobIdMap.getOrDefault(parameters.writeOptions.requestId, 0)
+        blobIdMap.put(parameters.writeOptions.requestId, blobIndexInBatch + 1)
+        KDSU.logInfo(
+          className,
+          s"Sealing blob in partition $partitionIdString for requestId: '${parameters.writeOptions.requestId}', " +
+            s"blob number ${row._2}, with size $count")
+        finalizeBlobWrite(blobWriter)
+        ingest(
+          blobWriter,
+          parameters,
+          ingestionProperties,
+          flushImmediately = !parameters.writeOptions.disableFlushImmediately,
+          curBlobUUID,
+          kustoClient,
+          partitionsResults,
+          batchIdForTracing,
+          blobIndexInBatch)
+        curBlobUUID = UUID.randomUUID().toString
+        createBlobWriter(
+          parameters.coordinates,
+          parameters.tmpTableName,
+          kustoClient,
+          partitionIdString,
+          row._2,
+          curBlobUUID)
+      }
     }
-
     KDSU.logInfo(
       className,
-      s"Finished serializing rows in partition $partitionIdString for " +
-        s"requestId: '${parameters.writeOptions.requestId}' ")
+      s"Finished serializing rows in partition $partitionIdString for requestId:'${parameters.writeOptions.requestId}'")
     finalizeBlobWrite(lastBlobWriter)
     if (lastBlobWriter.csvWriter.getCounter > 0) {
+      val blobIndexInBatch = blobIdMap.getOrDefault(parameters.writeOptions.requestId, 0)
+      blobIdMap.put(parameters.writeOptions.requestId, blobIndexInBatch + 1)
       ingest(
         lastBlobWriter,
-        lastBlobWriter.csvWriter.getCounter,
-        lastBlobWriter.sas,
+        parameters,
+        ingestionProperties,
         flushImmediately = false,
         curBlobUUID,
-        kustoClient)
+        kustoClient,
+        partitionsResults,
+        batchIdForTracing,
+        blobIndexInBatch)
+    }
+  }
+
+  def ingest(
+      blobResource: BlobWriteResource,
+      parameters: KustoWriteResource,
+      ingestionProperties: IngestionProperties,
+      flushImmediately: Boolean = false,
+      blobUUID: String,
+      kustoClient: ExtendedKustoClient,
+      partitionsResults: CollectionAccumulator[PartitionResult],
+      batchIdForTracing: String,
```

Comment:
use the batchIdForTracing as well in getDedupTagsPrefix - so that it will work on (spark) streaming as well.